primeqa/tableqa/tapex/tapex_component.py

```diff
@@ -0,0 +1,383 @@
+import numpy as np
+import pandas as pd
+from datasets import load_dataset
+
+from pickle import NONE
+from primeqa.tableqa.tapex.utils.argument_utils_for_tapex import DataTrainingArguments,ModelArguments
+import pandas as pd
+import transformers
+from filelock import FileLock
+import os
+from transformers import (
+    AutoConfig,
+    BartForConditionalGeneration,
+    DataCollatorForSeq2Seq,
+    HfArgumentParser,
+    Seq2SeqTrainer,
+    Seq2SeqTrainingArguments,
+    TapexTokenizer,
+    set_seed,
+)
+from transformers.file_utils import is_offline_mode
+from transformers.trainer_utils import get_last_checkpoint, is_main_process
+from transformers.utils import check_min_version
+from primeqa.tableqa.tapex.preprocessors.wikisql import preprocess_tableqa_function_wikisql
+from primeqa.tableqa.tapex.preprocessors.wikitablequestions import preprocess_tableqa_function_wtq
+from primeqa.tableqa.tapex.run_tapex import ModelArguments, DataTrainingArguments
+from primeqa.tableqa.tapex.metrics.tapex_accuracy import TapexAccuracy
+from primeqa.components.base import Reader
+
+logger = logging.getLogger(__name__)
+
+@dataclass
+class TapexReader(Reader):
+    def __init__(self,path_to_config_json): 
+        print("reading the config from ",path_to_config_json)
+        self._config_json = path_to_config_json
+
+    @property
+    def model(self):
+        """ Propery of TableQA model.
+        Returns:
+            Sequence to sequence model object (based on model name)
+        """
+        return self._model
+
+    @property
+    def tokenizer(self):
+        """ Property of Tapex model.
+        Returns:
+            Tokenizer class object based on the model name/ path
+        """
+        return self._tokenizer
+        
+    def __hash__(self):
+        class_name = 'TapexReader'
+        return hash(class_name)
+        
+        
+
+    def predict(self,data_dict,queries_list):
+        """This function takes a table dictionary and a list of queries as input and returns the answer to the queries using the TableQA model.
+
+        Args:
+            data_dict (Dict): Table in dict format
+            queries_list (List): List of queries
+
+        Returns:
+            Dict: Returns a dictionary of query and the predicted answer.
+        """
+        print("in predict for TapexModel with data: ",data_dict , " ,queries:", queries_list)
+        self.load(self._config_json)
+        table = pd.DataFrame.from_dict(data_dict)
+        inputs = self._tokenizer(table, queries_list, padding='max_length', return_tensors="pt")
+        outputs = self._model.generate(**inputs)
+        answers = self._tokenizer.batch_decode(outputs, skip_special_tokens=True)
+
+        query_answer_dict = {}
+        for query, answer in zip(queries_list, answers):
+            query_answer_dict[query] = answer
+        return query_answer_dict
+
+
+    def load(self,config_json) :
+        print("loading from config at ",config_json)
```

Comment:
Please loging instead of print.

Reply:
Fixed in 981f76a74229d8c41013ebec4d2287492e427a75

Reply:
> One minor point is that in other directories (mrc/tableqa), the `preprocessors` folder is inside a general `processors` directory. Right now TAPEX does not have postprocessors (only a very small `postprocess_text()` function in `run_tapex.py`), but in the future if dedicated postprocessors are needed (to accommodate new datasets, for example), it might be a good idea to push the current `preprocessors` (and any future `postprocessors`) inside a new directory `processors` and make things uniform with other directories.
==> introduced a processors/ folder to be compatible with mrc folder convention. Fixed in 981f76a74229d8c41013ebec4d2287492e427a75

Reply:
> Looks good. The code organization is easy to follow and the README is useful and thorough (points to notebooks to run a quick example application of TAPEX, for example).
> 
> One minor point is that in other directories (mrc/tableqa), the `preprocessors` folder is inside a general `processors` directory. Right now TAPEX does not have postprocessors (only a very small `postprocess_text()` function in `run_tapex.py`), but in the future if dedicated postprocessors are needed (to accommodate new datasets, for example), it might be a good idea to push the current `preprocessors` (and any future `postprocessors`) inside a new directory `processors` and make things uniform with other directories.

==> introduced a processors/ folder to be compatible with mrc folder convention. Fixed in https://github.com/primeqa/primeqa/commit/981f76a74229d8c41013ebec4d2287492e427a75

Reply:
> 1. I would prefer if the pre-processors were wrapped in a class rather than having some function in a file, to be more consistent with the pre-processors in MRC.
> 
> 
>     * primeqa/tableqa/tapex/preprocessors/wikisql.py
> 
>     * primeqa/tableqa/tapex/preprocessors/wikitablequestions.py
> 
> 
>     2. I was expecting run_tapex.py to be simplified by a lot by calling the init, predict, train methods from tapex_component.py.
>        The run scripts appear to be as complex as when we started out.
==> 1 is partially corrected by introducing a /processors folder aligned with primeqa/mrc folder structure, didn't introduce a new class becaue right now the usage is only about a function without any additional need.  
==> We can come back to handle suggestion 2 later, it is not affecting the functionality/usability.